esc_calibration: document latest functionality (#2599)
PX4/PX4-user_guide: en/advanced_config/esc_calibration.md

:::note
These instructions are only relevant to [PWM ESCs](../peripherals/pwm_escs_and_servo.md) and [OneShot ESCs](../peripherals/oneshot.md).
[DShot](../peripherals/dshot.md) and [CAN](../can/README.md) ESCs ([DroneCAN](../dronecan/escs.md)/Cyphal) do not require this kind of calibration.
:::

Electronic Speed Controllers (ESCs) regulate motor speed (and direction) based on the PWM input value from the flight controller (FC).
The range of inputs to which an ESC will respond is configurable, and the default range can differ even between ESCs of the same model.

This calibration updates all the ESCs with a fixed maximum (2000us) and minimum (1000us) PWM input from the flight controller.
Subsequently all the ESCs/motors will respond to flight controller input in the same way across the whole input range.

Calibration is recommended for all PWM/OneShot ESCs that support it, and in particular for low cost models.

:::note
High-quality controllers come with a factory calibration and should all respond the same way.

> In practice the input range may differ even on high quality controllers because you cannot be sure if they were manually recalibrated before and if that was done correctly. Also once calibrated they might not respond the same way as from the factory anymore.

Reviewer: ```suggestion
In practice the input range may differ even on high quality controllers because they may be have been manually recalibrated after leaving the factory, and that might not have been done correctly. Following manual calibration they might not respond in the same way as they did after factory calibration.

```

This calibration updates all the ESCs with a fixed maximum (2000us) and minimum (1000us) PWM input from the flight controller.
Subsequently all the ESCs/motors will respond to flight controller input in the same way across the whole input range.

Calibration is recommended for all PWM/OneShot ESCs that support it, and in particular for low cost models.

:::note
High-quality controllers come with a factory calibration and should all respond the same way.
In practice the input range may differ even on high quality controllers because they may be have been manually recalibrated after leaving the factory, and that might not have been done correctly. Following manual calibration they might not respond in the same way as they did after factory calibration.

Using this calibration tool is recommended, as it ensures that all ESCs behave exactly the same way.
:::

:::note
Some ESCs do not support such a calibration. They must come factory calibrated and should respond consistently out of the box.
This has to be verified using [Actuator Testing](../config/actuators.md#actuator-testing). Jump to step 7 which is still important.
:::

> OneShot ESCs should be [configured to use OneShot](../peripherals/oneshot.md#px4-configuration) before calibration otherwise calibration steps should be repeated after switching.

Reviewer: ```suggestion
OneShot ESCs should be [configured to use OneShot](../peripherals/oneshot.md#px4-configuration) before calibration (you should calibrate the ESCs after switching, even if you have previously calibrated).

```

## Preconditions

> You must be able to (un)power ESCs and flight controller independently. If the flight controller is powered through USB and there is a power module the procedure can detect battery connection. If battery power can't be detected e.g. because there's no power module then the calibration sequence is still performed based on timeouts.

Reviewer: ```suggestion
You must be able to (un)power ESCs and flight controller independently.
If the flight controller is powered through USB and there is a power module the procedure can detect battery connection.
If battery power can't be detected e.g. because there's no power module then the calibration sequence is still performed based on timeouts.

```

1. Remove the propellers.

   :::warning
   Never attempt ESC calibration with propellers on!

   The motors should not spin during ESC calibration.

>    However if the calibration starts when the ESC is powered already or the ESC doesn't properly support/detect the calibration sequence then it will respond to the PWM input by running the motor at maximum speed.

Reviewer: ```suggestion
   However if calibration starts when the ESC is powered already, or if the ESC doesn't properly support/detect the calibration sequence, then it will respond to the PWM input by running the motor at maximum speed.

```

   :::

> 1. Map the ESCs you're calibrating as Motors in the vehicle's [Actuator Configuration](../config/actuators.md). Only mapped actuators get an output and hence only ESCs mapped as motors get calibrated.

Reviewer: ```suggestion
1. Map the ESCs you're calibrating as Motors in the vehicle's [Actuator Configuration](../config/actuators.md).
   Only mapped actuators get an output, and hence only ESCs mapped as motors get calibrated.

```

> 1. Unpower the ESCs by unplugging the battery but keep the flight controller powered and connected to the ground station best by plugging it via USB (only).

Reviewer: ```suggestion
1. Unpower the ESCs by unplugging the battery.
   The flight controller should ideally still be powered via USB, and remain connected to the ground station.

```

Author: > flight controller should ideally still be powered

Maybe that's already covered but the flight controller must stay powered. That's easiest (= ideal?) achieved with the classic pixhawk boards because they stay powered with USB only.

For e.g. Skynode USB alone is not enough, you need to keep the power module connected and so it's more tricky. I'm not sure but wouldn't be surprised if there are other setups that work this way. Hence it's also mentioned in the troubleshooting. Hopefully, I can improve the workflow for these cases.

Reviewer: I will tidy this bit.

1. Open the *QGroundControl* **Settings > Power**, then press the **Calibrate** button.

   ![ESC Calibration step 1](../../assets/qgc/setup/esc/qgc_esc_calibration.png)

> 1. After successfully starting the calibration immediately power the ESCs:

Reviewer: ```suggestion
1. After starting the calibration sequence, power the ESCs (you should be prompted):

```

   ![ESC Calibration step 2](../../assets/qgc/setup/esc/esc_calibration_step_2.png)

   The calibration will begin automatically:

   ![ESC Calibration step 3](../../assets/qgc/setup/esc/esc_calibration_step_3.png)

> 1. You're informed once the calibration sequence finished. During the calibration you should hear model specific beeping from the ESC that indicates the individual steps of the calibration.

Reviewer: ```suggestion
1. During the calibration you should hear model specific beeping from the ESC, which indicates the individual steps of the calibration.

   You will be prompted when the calibration completes. 

```

   ![ESC Calibration step 4](../../assets/qgc/setup/esc/esc_calibration_step_4.png)

> 1. Go back to the [Actuator Configuration](../config/actuators.md). All motors with the same (re)calibrated ESCs should behave the same and work with the same configuration. Verify the following things:
>    - The minimum value for a motor (default: 1100us, parameters: [PWM_MAIN_MINn](../advanced_config/parameter_reference.md#PWM_MAIN_MIN1)/[PWM_AUX_MINn](../advanced_config/parameter_reference.md#PWM_AUX_MIN1)) should make the motor spin slowly but reliably and also spin up reliably after it was stopped. Use a bit of margin here to make sure it doesn't stop in air when it shouldn't.
>    To find the right value you can (still without propellers) test by changing the disarmed value e.g. to 1100us and 1050us and 1025us which makes the motor spin immediately and when you found the right value reset the disarmed value. After setting the desired minimum value you can test it using [Actuator Testing](../config/actuators.md#actuator-testing) by moving the test output slider just above the initial bump.
>    - The maximum value for a motor (default: 1900us, parameters: [PWM_MAIN_MAXn](../advanced_config/parameter_reference.md#PWM_MAIN_MAX1)/[PWM_AUX_MAXn](../advanced_config/parameter_reference.md#PWM_AUX_MAX1)) should be chosen such that if the value is further increased from there the motor doesn't actually spin faster anymore.
>    This can again be done by changing the disarmed value e.g. to 1900us and 1950us and 1975us and observing e.g. audibly if the motor is still going faster.
>    - The disarmed value for a motor (default: 1000us, parameters: [PWM_MAIN_DISn](../advanced_config/parameter_reference.md#PWM_MAIN_DIS1)/[PWM_AUX_DISn](../advanced_config/parameter_reference.md#PWM_AUX_DIS1)) should make the motor stop and stay stopped. When calibrated properly this value should be able to stay 1000us. If the ESC was previously miscalibrated then the motor might already spin with 1000us.

Reviewer: ```suggestion
1. Go back to the [Actuator Configuration](../config/actuators.md).
   All motors with the same (re)calibrated ESCs should behave the same and work with the same configuration.
   Verify the following things:
   
   - The minimum value for a motor (default: 1100us, parameters: [PWM_MAIN_MINn](../advanced_config/parameter_reference.md#PWM_MAIN_MIN1)/[PWM_AUX_MINn](../advanced_config/parameter_reference.md#PWM_AUX_MIN1)) should make the motor spin slowly but reliably and also spin up reliably after it was stopped.
     Use a bit of margin here to make sure it doesn't stop in air when it shouldn't.
     To find the right value you can (still without propellers) test by changing the disarmed value e.g. to 1100us and 1050us and 1025us which makes the motor spin immediately and when you found the right value reset the disarmed value.
     After setting the desired minimum value you can test it using [Actuator Testing](../config/actuators.md#actuator-testing) by moving the test output slider just above the initial bump.
   - The maximum value for a motor (default: 1900us, parameters: [PWM_MAIN_MAXn](../advanced_config/parameter_reference.md#PWM_MAIN_MAX1)/[PWM_AUX_MAXn](../advanced_config/parameter_reference.md#PWM_AUX_MAX1)) should be chosen such that if the value is further increased from there the motor doesn't actually spin faster anymore.
     This can again be done by changing the disarmed value e.g. to 1900us and 1950us and 1975us and observing e.g. audibly if the motor is still going faster.
   - The disarmed value for a motor (default: 1000us, parameters: [PWM_MAIN_DISn](../advanced_config/parameter_reference.md#PWM_MAIN_DIS1)/[PWM_AUX_DISn](../advanced_config/parameter_reference.md#PWM_AUX_DIS1)) should make the motor stop and stay stopped.
     When calibrated properly this value should be able to stay 1000us.
     If the ESC was previously mis-calibrated then the motor might already spin with 1000us.

```

   :::note

>    VTOL or fixed-wing motors do not need any special PWM configuration. They will automatically stop during flight when desired by the autopilot with the PWM configuration as described.

Reviewer: @MaEtUgR Is this more correct? Most of my changes in this update are layout only, but this one is a content change.

```suggestion
   VTOL and fixed-wing motors do not need any special PWM configuration.
   They will automatically stop during flight when commanded by the autopilot with the default PWM configuration.

```

Author: Completely correct. This is a huge help!

   :::

## Troubleshooting

> Possible problems:
> 1. If you do not plug any ESC at the right time or the ESCs don't support calibration you'll still get the message that the calibration successfully completed because the system cannot possibly know if it actually worked.
> 1. If you have a power module configured and connected the system will disallow starting the calibration for safety reasons. Unplug power to the ESCs first. If you're blocked because a power module is necessary to keep your flight controller alive but can (un)power the ESCs separately make sure the power module is disabled in PX4 just for the ESC calibration e.g. by [PWM_AUX_MINn](../advanced_config/parameter_reference.md#PWM_AUX_MIN1).
> 1. If there is a power module and the system detects an increase in current consumption directly after successfully initiating calibration it will be aborted for safety reasons.

Reviewer: ```suggestion

Possible problems:

1. If you do not plug any ESC at the right time or the ESCs don't support calibration you'll still get the message that the calibration successfully completed because the system cannot possibly know if it actually worked.
1. If you have a power module configured and connected the system will disallow starting the calibration for safety reasons.
   Unplug power to the ESCs first. If you're blocked because a power module is necessary to keep your flight controller alive but can (un)power the ESCs separately make sure the power module is disabled in PX4 just for the ESC calibration e.g. by [PWM_AUX_MINn](../advanced_config/parameter_reference.md#PWM_AUX_MIN1).
1. If there is a power module and the system detects an increase in current consumption directly after successfully initiating calibration it will be aborted for safety reasons.

```

   :::
1. Map the ESCs you're calibrating as motors in the vehicle's [Actuator Configuration](../config/actuators.md).
   Only mapped actuators get an output, and hence only ESCs mapped as motors will be calibrated.

1. Unpower the ESCs by unplugging the battery.
   The flight controller should still be powered via USB, and remain connected to the ground station.

1. Open the *QGroundControl* **Settings > Power**, then press the **Calibrate** button.

   ![ESC Calibration step 1](../../assets/qgc/setup/esc/qgc_esc_calibration.png)

1. After starting the calibration sequence, power the ESCs (you should be prompted):

   ![ESC Calibration step 2](../../assets/qgc/setup/esc/esc_calibration_step_2.png)

   The calibration will begin automatically:

   ![ESC Calibration step 3](../../assets/qgc/setup/esc/esc_calibration_step_3.png)

1. During the calibration you should hear model-specific beeping from the ESC, which indicates the individual steps of the calibration.

   You will be prompted when the calibration completes.

   <a id="actuatorconfig_step"></a>
   ![ESC Calibration step 4](../../assets/qgc/setup/esc/esc_calibration_step_4.png)

> 1. Go back to the [Actuator Configuration](../config/actuators.md) section.

Reviewer: I've modified this a bit to remove the parameter names (you shouldn't need to know them when using the actuators config). I've also modified the steps so you use the slider in the current position, rather than messing with the disarmed value and having to set it back afterwards. I just think this is more logical.

NOTE, http://docs.px4.io/main/en/config/actuators.html#actuator-configuration-and-testing is the same process. This has two implications:

1. We should fix up the images/defaults in http://docs.px4.io/main/en/config/actuators.html#actuator-configuration-and-testing to note the new defaults.
2. Should we just link to this section rather than reproducing>"
3. Does http://docs.px4.io/main/en/config/actuators.html#actuator-configuration-and-testing need a bullet point for setting the _maximum_ value as you have here?

Author: You are finding the rough spots immediately 👍 

> you shouldn't need to know them when using the actuators config

Completely true for going through the procedure. I only mentioned them for completeness e.g. for people who want to take note and put these settings into an airframe configuration. Not sure if it's appropriate.

> so you use the slider in the current position, rather than messing with the disarmed value

That sounds more logical yes. But I prefer the more hacky-sounding way with the disarmed value. The goal is to find an explicit value by trial and observation and you might realistically end up trying 10 times including all motors. If you do that with the slider the workflow is:
- type a value in the minimum field
- turn the sliders on
- move the slider over the bump
- observe the result and repeat

I'm using the disarmed field because then it's just:
- type a value in the disarmed field
- observe the result and repeat

Additionally, I can't find the maximum value any other way because I do it by listening if the motor spins any faster while changing the command. With the slider workflow I cannot hear the direct difference because the motor gets stopped and spun up again in between tries.

> is the same process

There's especially overlap with https://docs.px4.io/main/en/config/actuators.html#motor-configuration I'm aware this is not ideal. My goal is to have people not skim through the large actuators page and miss the fact that correctly configuring the limits is crucial and in principle more important than the ESC calibration itself.

> Should we just link to this section rather than reproducing

Yes but what's the best way to do that? Making sure users with short attention span go through the important steps with the least amount of confusion. Probably my approach is too local.

> need a bullet point for setting the maximum value as you have here?

Yes, it's part of a bring-up. It usually gets neglected because it's not important for the happy path flying. But ESCs commonly also have a dead zone at the top of the command range. Not configuring it leads to worse performance in corner cases. The controller/allocation relies on that high thrust actuation range being available. If there's no authority for the motor to actually spin faster that's unexpectedly non-linear and tracking gets unpredictably worse whenever that command is hit during flight. Can for example result in wobbles and attitude overshoot when flying with full thrust.

Reviewer: You're most welcome!

> so you use the slider in the current position, rather than messing with the disarmed value

Thanks for the explanation. I have rewritten this as you would prefer in https://github.com/PX4/PX4-user_guide/pull/2599/commits/5b748488c917a07fcc2e1756fb8ffd0670b30433 - needs to be checked, and I have a question.

> My goal is to have people not skim through the large actuators page and miss the fact that correctly configuring the limits is crucial and in principle more important than the ESC calibration itself.

I'm fine with reproducing it here - the actuator doc is a monolith. I want to split it a bit, but will probably do so in a following release (a scary job). At that point maybe we revisit whether we want to duplicate this or not.

What we will need now is to update the Actuators doc to show the new defaults for all the images. I guess that will have to wait for your PR to go in?

> need a bullet point for setting the maximum value as you have here?

Great. I added one. Will add a separate discussion point there.

     You can confirm this in [Actuator Testing](../config/actuators.md#actuator-testing) by moving the test output slider to the snap position at the bottom of the slider and observing that the motor does not spin.

     If the ESC spins with the default value of 1000us then the ESC is not properly calibrated.
     If using an ESC that can't be calibrated, you should reduce the PWM output value for the output to below where the ESC does not split (such as 900us).

   :::note
   VTOL and fixed-wing motors do not need any special PWM configuration.
   With the default PWM configuration they will automatically stop during flight when commanded by the autopilot.
   :::

## Troubleshooting

1. Calibration can state that it has succeeded even when it has failed.

   This happens happen if you do not power the ESC at the right time, or the ESCs don't support calibration.
   This occurse because PX4 has has no way to know whether or not calibration was successful.

1. Calibration cannot be started if you have a power module configured and connected (for safety reasons).

   Unplug power to the ESCs first.

>    If you're blocked because a power module is necessary to keep your flight controller alive, but you can (un)power the ESCs separately, you can disable the motor just for the ESC calibration using parameters such as [PWM_AUX_MINn](../advanced_config/parameter_reference.md#PWM_AUX_MIN1).

Reviewer: Are you saying that in order to calibrate the motor must be unpowered, and you can do this by setting the minimum value to (all) motors to some low value - e.g. PWM_AUX_MIN1 to 900us?

FYI, I updated all these items to state "problem first". This usually a better approach.

Author: It's a copy paste mistake on my side. I copied the parameter link but didn't change to the correct name 🤦‍♂️ 
Thanks for pointing it out it only occurred to me now when reading your very helpful improvements.

     You can confirm that a motor spins at minimum (still without propellers) in [Actuator Testing](../config/actuators.md#actuator-testing), by enabling the sliders, and then moving the test output slider for the motor to the first snap position from the bottom.
     The correct value should make the motor spin immediately and reliably as you move the slider from disarmed to minimum.

     To find the "optimal" minimum value, move the slider to the bottom (disarmed).
     Then increase the PWM output's `disarmed` setting in small increments (e.g. 1025us, 1050us, etc), until the motor starts to spin reliably (it is better to be a little too high than a little too low).
     Enter this value into the `minimum` setting for all the motor PWM outputs, and restore the `disarmed` output to `1100us`.

   - The maximum value for a motor (default: `1900us`) should be chosen such that increasing the value doesn't make the motor spin any faster.

     You can confirm that the motor spins quickly at the maximum setting in [Actuator Testing](../config/actuators.md#actuator-testing), by moving the associated test output slider to the top snap position.

>      To find the "optimal" maximum value, first move the slider to the bottom (disarmed).

Reviewer: My assumption here is that props are still off - correct?

>      Then increase the PWM output's `disarmed` setting to near the default maximum (`1900`) - the motors should spin up.

Reviewer: Should we suggest a lower value as starting point - i.e. is it possible that the 1900 value will already be too high?

Reviewer: I will assume not.

1. Remove the propellers.

   :::warning
   Never attempt ESC calibration with propellers on!

   The motors _should_ not spin during ESC calibration.
   However if calibration starts when the ESC is already powered, or if the ESC doesn't properly support/detect the calibration sequence, then it will respond to the PWM input by running the motor at maximum speed.
   :::
1. Map the ESCs you're calibrating as motors in the vehicle's [Actuator Configuration](../config/actuators.md).
   Only mapped actuators get an output and only ESCs mapped as motors will be calibrated.

1. Unpower the ESCs by unplugging the battery.

>    The flight controller must stay powered e.g. through USB remaining connected to the ground station.

Reviewer: ```suggestion
   The flight controller must stay powered, for example, by keeping the USB connected to the ground station.

```
